internals-js/src/buildSchema.ts

```diff
@@ -171,17 +171,46 @@ function buildNamedTypeAndDirectivesShallow(documentNode: DocumentNode, schema:
       case 'UnionTypeDefinition':
       case 'EnumTypeDefinition':
       case 'InputObjectTypeDefinition':
+        let type = schema.type(definitionNode.name.value);
+        // Note that the type may already exists due to an extension having been processed first, but we know we
+        // have seen 2 definitions (which is invalid) if the definition has `preserverEmptyDefnition` already set
+        // since it's only set for definitions, not extensions. 
+        // Also note that we allow to redefine built-ins.
+        if (!type || type.isBuiltIn) {
+          type = schema.addType(newNamedType(withoutTrailingDefinition(definitionNode.kind), definitionNode.name.value));
+        } else if (type.preserveEmptyDefinition)  {
+          // Note: we reuse the same error message than graphQL-js would output
+          throw new GraphQLError(`There can be only one type named "${definitionNode.name.value}"`);
```

Comment:
Reviewer note: writing this, I realised that we weren't validating this properly. That is:
```graphql
type Foo {
  bar: Int
}

type Foo {
  baz: Int
}
```
was not erroring out, even though it should as it's not valid graphQL. So took the liberty to fix as well (and there is a test for it).